src/cong.cc

```diff
@@ -241,7 +240,60 @@ namespace libsemigroups {
     }
   }
 
+  bool Congruence::is_obviously_infinite() {
+    // We apply some simple, quick checks that may establish that the congruence
+    // has an infinite number of classes; if so, Todd-Coxeter should not be run.
+    // If this returns false, it is safe to run Todd-Coxeter, although of course
+    // Todd-Coxeter may still run forever.
+
+    // If we have a concrete semigroup, it should be finite.
+    if (_semigroup != nullptr) {
+      return false;
+    }
+
+    // If there are no rels, or more gens than rels, it must be infinite
+    if (_nrgens > _relations.size() + _extra.size()) {
+      return true;
+    }
+
+    // Does there exist a generator which appears in no relation?
+    bool found;
+    for (size_t gen = 0; gen < _nrgens; gen++) {
+      found = false;
+      for (const relation_t& rel : _relations) {
+        if (std::find(rel.first.cbegin(), rel.first.cend(), gen)
+            != rel.first.cend()) {
```

Comment:
You could combine this line with the next `if`:

    if (std::find(rel.first.cbegin(), rel.first.cend(), gen)
             != rel.first.cend() || std::find(rel.first.cbegin(), rel.first.cend(), gen) != rel.second.cend()) {